src/values/AbstractValue.js

```diff
@@ -389,13 +434,58 @@ export default class AbstractValue extends Value {
     return result;
   }
 
+  static createFromLogicalOp(
+    realm: Realm,
+    op: BabelNodeLogicalOperator,
+    left: Value,
+    right: Value,
+    loc?: ?BabelNodeSourceLocation
+  ): AbstractValue {
+    let leftTypes, leftValues;
+    if (left instanceof AbstractValue) {
+      leftTypes = left.types;
+      leftValues = left.values;
+    } else {
+      leftTypes = new TypesDomain(left.getType());
+      invariant(left instanceof ConcreteValue);
+      leftValues = new ValuesDomain(left);
+    }
+
+    let rightTypes, rightValues;
+    if (right instanceof AbstractValue) {
+      rightTypes = right.types;
+      rightValues = right.values;
+    } else {
+      rightTypes = new TypesDomain(right.getType());
+      invariant(right instanceof ConcreteValue);
+      rightValues = new ValuesDomain(right);
+    }
+
+    let resultTypes = TypesDomain.logicalOp(op, leftTypes, rightTypes);
+    let resultValues = ValuesDomain.logicalOp(realm, op, leftValues, rightValues);
+    let [hash, args] = hashCall(op, left, right);
+    let Constructor = Value.isTypeCompatibleWith(resultTypes.getType(), ObjectValue)
+      ? AbstractObjectValue
+      : AbstractValue;
+    let result = new Constructor(realm, resultTypes, resultValues, hash, args, ([x, y]) =>
+      t.logicalExpression(op, x, y)
+    );
+    result.kind = op;
+    result.expressionLocation = loc;
+    return result;
+  }
+
   static createFromConditionalOp(
     realm: Realm,
-    condition: Value,
+    condition: AbstractValue,
     left: void | Value,
     right: void | Value,
     loc?: ?BabelNodeSourceLocation
   ): AbstractValue {
+    if (left instanceof BooleanValue && right instanceof BooleanValue) {
```

Comment:
Why special casing for `BooleanValue` and not just using the more general `.mightNotBeTrue` kind of machinery?

Reply:
This is more specific than it needs to be. I'll generalize it, but I can't really think of a valid use case for the more general form.